### DISCUSSION TITLE
What do the uv__run_idle and uv__run_prepare operations do in node.js?

### DISCUSSION THREAD

**Discussion by dSalieri** [libuv/libuv#4462] [Q&A] [answered]
I've read a lot of articles online and also read the [node.js documentation](https://nodejs.org/en/learn/asynchronous-work/event-loop-timers-and-nexttick) and also read an [overview of the libuv](https://docs.libuv.org/en/v1.x/design.html).

> uv__run_idle (cycle)
> uv__run_prepare (cycle)

What says libuv:
> 5. Idle handle callbacks are called. Despite the unfortunate name, idle handles are run on every loop iteration, if they are active.
>6. Prepare handle callbacks are called. Prepare handles get their callbacks called right before the loop will block for I/O.

Can idle callbacks and prepare callbacks be created via javascript on node.js platform? If not, then why are they and what do they affect?

What node.js doc and lot of articles say:
> **idle, prepare:** [only used internally](https://nodejs.org/en/learn/asynchronous-work/event-loop-timers-and-nexttick#phases-overview). 

And what does it mean? That is, on the node.js platform we cannot reach these pointed stages in any way?

Then there are two questions:
1) What do these operations actually do? (their definition is similar to that of `uv__run_check(loop)` and this stage is somewhere tied to `setImmediate` processing but I don’t know where).
2) If you remove `uv__run_idle(loop)` and `uv__run_prepare(loop)` from the node.js codebase, will something break or what will this affect?

I still haven't seen answers to these questions, which is very strange because the (**idle, prepare**) phase in the node.js documentation is quoted from every corner, but no one knows what this phase does, which is sad.

P.S. I looked at the source code in libuv, but it didn’t give me anything, because you have to know the structure of node.js and its components.

**Comment by bnoordhuis** [comment, marked as answer]
You're asking in the wrong place for your node.js-related questions. I'll answer these two though:

> What do these operations actually do?

Idle: the event loop normally blocks waiting for I/O or a timer to expire, whichever happens first. uv_idle_start() makes it not block. uv_spin_start() would have been a better name in retrospect but :shrug:

Prepare: called before libuv polls for I/O; check handles are called afterwards. They're useful for when you have e.g. a mutex that you want to release while waiting and reacquire when woken up.

> on the node.js platform we cannot reach these pointed stages in any way?

AFAIK, it doesn't expose them, no.

**Reply by dSalieri** [reply]
@bnoordhuis this explains a lot. I discovered that the idle phase in node.js is needed in order not to block I/O in the poll phase and almost immediately go to the check stage to execute setImmediate callbacks. The prepare phase is associated with [StartProfilerIdleNotifier](https://github.com/nodejs/node/blob/1eac9a3d41b304c0dad86b1d8188a893afb7a75d/src/env.cc#L1197), but what it is is unclear, but it seems that it is for profiling.

Thanks for the answer.

P.S I asked the same question in the node.js repository and no one answered, it was not in vain that I asked here.

**Reply by bnoordhuis** [reply]
The "idle" in StartProfilerIdleNotifier is unrelated to uv_idle_t.

It tells V8 that samples recorded by the CPU profiler should be marked in a way that post-processing tools can filter them out, to avoid traces where 99.9% is spent inside epoll_wait(), drowning out everything else.

**Reply by dSalieri** [reply]
@bnoordhuis that is, the chrome profiler gets already filtered data? How do I know that the trace part refers to calls inside epoll_wait()? Are epoll_wait() traces wait/idle records or something like that?

**Reply by bnoordhuis** [reply]
I wrote it specifically for the tick-processor cli tool (what is now the `--prof-process` flag); devtools uses a different mechanism.

> Are epoll_wait() traces wait/idle records or something like that?

Yes, IIUC your question. The CPU profiler records a sample every millisecond. If your program is doing nothing most of the time, then most of those samples get attributed to epoll_wait(), because that's what node/libuv uses to suspend itself and wait for I/O.

**Reply by dSalieri** [reply]
@bnoordhuis small clarification, when StartProfilerIdleNotifier is used, which records are marked for filtering: epoll_wait() or all the others? Just curious to know.

**Reply by bnoordhuis** [reply]
Everything until idle mode is turned off again. Node basically calls `isolate->SetIdle(true)` in the prepare callback and `isolate->SetIdle(false)` in the check callback.

Honestly, it's probably not the best solution anymore because CPU time for I/O callbacks and a bunch of other things gets misattributed, but I won't be working on that unless someone pays me. :-)

**Reply by dSalieri** [reply]
@bnoordhuis oh, sorry for overusing comments, could you look at [this](https://github.com/orgs/nodejs/discussions/54037) or [this](https://github.com/libuv/libuv/discussions/4466) question (actually they are same but formulated slightly differently)? I'm sure you know the answer. I don't want to miss the opportunity while you're answering.

I did everything to find out the answer to those questions, but no one can answer, because only experts like you know the answer.
